fil_fungible_token/src/token/mod.rs

```diff
@@ -278,8 +285,380 @@ where
+        token
+            .transfer(spender, owner, receiver, TokenAmount::from(60))
+            .unwrap_err();
+
+        // because the allowance is only 10
+        let allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(allowance, TokenAmount::from(10));
+
+        // spender can transfer 1
+        token
+            .transfer(spender, owner, receiver, TokenAmount::from(1))
+            .unwrap();
+
+        let allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(allowance, TokenAmount::from(9));
+
+        // owner revokes the rest
+        token.revoke_allowance(owner, spender).unwrap();
+
+        let allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(allowance, TokenAmount::from(0));
+
+        // spender can no longer transfer 1
+        token
+            .transfer(spender, owner, receiver, TokenAmount::from(1))
+            .unwrap_err();
+
+        // only the 1 token transfer should have succeeded
+        // verify all balances are correct
+        let owner_balance = token.balance_of(owner).unwrap();
+        let receiver_balance = token.balance_of(receiver).unwrap();
+        let spender_balance = token.balance_of(spender).unwrap();
+        assert_eq!(owner_balance, TokenAmount::from(99));
+        assert_eq!(receiver_balance, TokenAmount::from(1));
+        assert_eq!(spender_balance, TokenAmount::from(0));
+    }
+
+    #[test]
+    fn it_disallows_transfer_when_insufficient_allowance() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let owner = 1;
+        let receiver = 2;
+        let spender = 3;
+
+        // mint 100 for the owner
+        token.mint(owner, TokenAmount::from(100)).unwrap();
+        // approve only 40 spending allowance for spender
+        token
+            .increase_allowance(owner, spender, TokenAmount::from(40))
+            .unwrap();
+        // spender attempts makes transfer of 60 from owner -> receiver
+        // this is within the owner's balance but not within the spender's allowance
+        token
+            .transfer(spender, owner, receiver, TokenAmount::from(60))
+            .unwrap_err();
+
+        // verify all balances are correct
+        let owner_balance = token.balance_of(owner).unwrap();
+        let receiver_balance = token.balance_of(receiver).unwrap();
+        let spender_balance = token.balance_of(spender).unwrap();
+        assert_eq!(owner_balance, TokenAmount::from(100));
+        assert_eq!(receiver_balance, TokenAmount::zero());
+        assert_eq!(spender_balance, TokenAmount::zero());
+
+        // verify allowance was not spent
+        let spender_allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(spender_allowance, TokenAmount::from(40));
+    }
+
+    #[test]
+    fn it_checks_for_invalid_negatives() {
```

Comment:
Ah, I see. I think better to roll these in with the other tests for each functionality.